queue/src/main/java/org/killbill/queue/DefaultQueueLifecycle.java

```diff
@@ -82,66 +103,90 @@ public void run() {
+                    completeOrRetryProcessedEvents();
                     log.info(String.format("%s: Thread %s has exited", svcQName, Thread.currentThread().getName()));
                 }
             }
 
+            private void dispatchEvents() {
+                long ini = System.nanoTime();
+                final DispatchResultMetrics metricsResult = doDispatchEvents();
+                dispatchedEntries.update(metricsResult.getNbEntries());
+                if (isStickyEvent) {
+                    dispatchTime.update(metricsResult.getTimeNanoSec(), TimeUnit.NANOSECONDS);
+                } else {
+                    dispatchTime.update(System.nanoTime() - ini, TimeUnit.NANOSECONDS);
+                }
+            }
+
+            private void completeOrRetryProcessedEvents() {
+                long ini = System.nanoTime();
+                int completed = drainCompletedEvents();
+                int retried = drainRetriedEvents();
+                final int completeOrRetried = completed + retried;
+                if (completeOrRetried > 0) {
+                    completeEntries.update(completeOrRetried);
+                    completeTime.update(System.nanoTime() - ini, TimeUnit.NANOSECONDS);
+                }
+            }
+
             // Move completed entries through batches using the same main DB (lifecycle) thread
-            private void drainCompletedEvents() {
+            private int drainCompletedEvents() {
                 int curSize = completedOrFailedEvents.size();
                 if (curSize > 0) {
                     final List<EventEntryModelDao> completed = new ArrayList<>(curSize);
                     completedOrFailedEvents.drainTo(completed, curSize);
                     doProcessCompletedEvents(completed);
                 }
+                return curSize;
             }
 
             // Move retried entries using the same main DB (lifecycle) thread
-            private void drainRetriedEvents() {
+            private int drainRetriedEvents() {
                 final int curSize = retriedEvents.size();
                 if (curSize > 0) {
                     final List<EventEntryModelDao> retried = new ArrayList<>(curSize);
                     retriedEvents.drainTo(retried, curSize);
                     doProcessRetriedEvents(retried);
                 }
+                return curSize;
             }
 
-            private void sleepALittle(final long loopTimeMsec) throws InterruptedException {
-                if (config.getPersistentQueueMode() == PersistentQueueConfig.PersistentQueueMode.STICKY_EVENTS) {
-                    // Disregard config.getPollingSleepTimeMs() in that mode in case this is not correctly configured with 0
+            private void sleepSporadically(final long loopTimeMsec) throws InterruptedException {
+                if (isStickyEvent) {
+                    // In this mode, the main thread does not sleep, but blocks on the inflightQ to minimize latency.
```

Comment:
Thought: could we change the code to be always in sticky events?

1. Get rid of the polling mode for bus
2. For notifications, have a separate thread that fetches the ready events and put them in the "inflight" queue, for the main dispatcher thread to pull from

I don't think it would change the overall behavior, but it could simplify the code a bit further?

Reply:
Let's talk..